include/revng/ADT/GenericGraph.h

```diff
@@ -1463,4 +1505,49 @@ struct GraphTraits<llvm::Inverse<T *>>
   static size_t size(T *G) { return G->size(); }
 };
 
+// Because of what seems to be a clang bug (it works fine in both gcc and msvc),
+// the deduction of qualified types restricted by a requires clause in a type
+// specialization is not possible, so to make it work until the bug is fixed,
+// we need to shift the partial type template deduction into a function template
+// deduction, hence this trick:
+template<SpecializationOfForwardNode T>
+consteval auto forwardsSpecializationTrick() {
+  return ForwardsNodeTraitsImpl<T>{};
+}
+template<SpecializationOfMutableEdgeNode T>
+consteval auto forwardsSpecializationTrick() {
+  return MutableEdgeNodeTraitsImpl<T>{};
+}
+template<SpecializationOfGenericGraph T>
+consteval auto forwardsSpecializationTrick() {
+  return GraphTraitsImpl<T>{};
+}
+
+template<SpecializationOfBidirectionalNode T>
+consteval auto backwardsSpecializationTrick() {
+  return BackwardsNodeTraitsImpl<T>{};
+}
+template<SpecializationOfMutableEdgeNode T>
+consteval auto backwardsSpecializationTrick() {
+  return InverseMutableEdgeNodeTraitsImpl<T>{};
+}
+template<SpecializationOfGenericGraph T>
+consteval auto backwardsSpecializationTrick() {
+  return InverseGraphTraitsImpl<T>{};
+}
+
+// TODO: remove this and declare the traits directly once the bug is fixed.
+
+} // namespace revng::detail
+
+namespace llvm {
+
+template<typename T>
+struct GraphTraits<T *>
+  : public decltype(revng::detail::forwardsSpecializationTrick<T>()) {};
+
+template<typename T>
+struct GraphTraits<Inverse<T *>>
+  : public decltype(revng::detail::backwardsSpecializationTrick<T>()) {};
+
```

Comment:
Mmmh, I'm not very happy with the final form of the changes to `GenericGraph.h`.

It became strictly more difficult and hackish to read and what do we get back?

Reply:
> what do we get back?

No need to keep the "SomethingSomethingTag" within the classes. Also, it's a bit more generic (these things can be used on unrelated classes as long as they are sufficiently `GenericGraph`-like).

But yeah, I get what you mean - the bug makes it quite ugly. Consider, though, that it's a temporary solution that would eventually disappear.

Reply:
Either way, I'm not opposed from stashing this in a branch somewhere and restoring this approach once we update to a clang version that doesn't suffer from this anymore.

Reply:
This PR has been open too long, let's try to get it down to non-controversial stuff.

Reply:
Sounds good.

Reply:
Stashed it into here for now:
https://github.com/Cvelth/revng/tree/stash/non-strict-generic-graph-concepts